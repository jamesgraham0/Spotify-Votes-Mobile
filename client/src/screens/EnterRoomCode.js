import { StyleSheet, Text, View, TouchableOpacity, TextInput, KeyboardAvoidingView, Keyboard } from 'react-native';
import React, { useState, useEffect } from 'react';
import { socket } from '../utils/socket';
import Ionicons from '@expo/vector-icons/Ionicons';
import Constants from '../utils/constants';
import DarkBackgroundCircles from '../components/DarkBackgroundCircles';

const EnterRoomCode = ({ navigation, route }) => {
    const [code, setCode] = useState(['', '', '', '', '']);
    const { user } = route.params;
    const inputRefs = [];

    useEffect(() => {
        if (code[code.length - 1] !== '') {
            const nextIndex = code.findIndex((digit) => digit === '');
            if (nextIndex !== -1) {
                inputRefs[nextIndex].focus();
            } else {
                validateCodeAndJoinRoom();
                inputRefs[0].focus();
            }
        }
    }, [code]);

    const validateCodeAndJoinRoom = async () => {
        Keyboard.dismiss();
        const enteredCode = code.join('');
        socket.emit('checkRoomCode', enteredCode, ({ room, isCorrectCode }) => {
            if (isCorrectCode) {
                socket.emit('joinRoom', { user: user, room: room });
                navigation.navigate('Room', { user: user, room: room });
            } else {
                inputRefs[0].focus();
                alert(`There is no room with code ${enteredCode}`);
                setCode(['', '', '', '', '']);
            }
        });
    };

    const handleReturnToJoinOrCreateRoom = () => {
        navigation.navigate('JoinOrCreateRoom', { user: user });
    }


    const onChanged = (text, index) => {
        const newCode = [...code];
        if (text.length > 0) {
            newCode[index] = text.charAt(text.length - 1);
            setCode(newCode);
            if (index < code.length - 1) {
                inputRefs[index + 1].focus();
            }
        }
        else {
            newCode[index] = '';
            setCode(newCode);
            if (index > 0) {
                inputRefs[index - 1].focus();
            }
        }
    };

    const handleBackspaceOnEmptyBox = (index) => {
        const newCode = [...code];
        if (index > 0) {
            newCode[index - 1] = '';
            setCode(newCode);
            inputRefs[index - 1].focus();
        }
    };

    const renderCodeInputs = () => {
        return (
            <View style={styles.codeInputContainer}>
                {code.map((digit, index) => (
                    <View key={index} style={styles.codeBox}>
                        <TextInput
                            style={styles.codeInput}
                            autoFocus={index === 0}
                            ref={(ref) => inputRefs[index] = ref}
                            keyboardType='numeric'
                            onChangeText={text => onChanged(text, index)}
                            onKeyPress={({ nativeEvent }) => {
                                if (nativeEvent.key === 'Backspace') {
                                  handleBackspaceOnEmptyBox(index);
                                };
                              }}
                            value={digit}
                            maxLength={1}
                        />
                    </View>
                ))}
            </View>
        );
    };

    const header = () => {
        return <View style={Constants.HEADER_STYLES}>
                    <TouchableOpacity
                        onPress={handleReturnToJoinOrCreateRoom}
                        style={styles.returnButton}
                    >
                        <Ionicons name="chevron-back-circle-outline" size={32} color="grey" />
                    </TouchableOpacity>
                    <Text style={styles.enterRoomCode}>Enter Room Code</Text>
                </View>
    };

    return (
        <View style={styles.outerContainer}>
            <DarkBackgroundCircles/>
            {header()}
            <Text style={styles.instructionText}>Enter the code for the room you want to join</Text>
            <KeyboardAvoidingView
                behavior={Platform.OS === "ios" ? "padding" : "height"}
                style={styles.writeTaskWrapper}
            >
                {renderCodeInputs()}
            </KeyboardAvoidingView>
        </View>
    )
}

const styles = StyleSheet.create({
    outerContainer: {
        flex: 1,
        backgroundColor: '#191414',
    },
    enterRoomCode: {
        color: '#fff',
        fontSize: 24,
        maxWidth: '70%',
        textAlign: 'center',
        marginHorizontal: 34,
        marginTop: 20,
    },
    returnButton: {
        width: 50,
        height: 50,
        marginTop: 36,
    },
    instructionText: {
        color: '#BBB',
        fontSize: 20,
        marginHorizontal: 60,
        marginTop: 100,
        marginBottom: 10,
    },
    writeTaskWrapper: {
        width: '100%',
        flexDirection: 'row',
        justifyContent: 'space-around',
        alignItems: 'center',
        marginVertical: 100,
    },
    codeInputContainer: {
        flexDirection: 'row',
        justifyContent: 'center',
        marginBottom: 200,
    },
    codeBox: {
        width: 60,
        height: 60,
        marginHorizontal: 5,
        backgroundColor: 'black',
        borderRadius: 10,
        borderWidth: 1,
        borderColor: 'grey',
    },
    codeInput: {
        flex: 1,
        color: 'white',
        fontSize: 24,
        textAlign: 'center',
    },
})


export default EnterRoomCode;