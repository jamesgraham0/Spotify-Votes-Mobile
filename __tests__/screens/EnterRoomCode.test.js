import React from 'react';
import renderer from 'react-test-renderer';
import EnterRoomCode from '../../client/src/screens/EnterRoomCode';

describe('EnterRoomCode', () => {
    it('should render the EnterRoomCode component', () => {
        const navigation = {};
        const route = { params: { user: {} } };
        const tree = renderer.create(<EnterRoomCode navigation={navigation} route={route} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});