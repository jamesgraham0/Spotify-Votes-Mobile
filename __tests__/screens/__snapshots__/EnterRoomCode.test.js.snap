// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnterRoomCode should render the EnterRoomCode component 1`] = `
<View
  style={
    {
      "backgroundColor": "#191414",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      style={
        [
          {
            "borderRadius": 100,
            "elevation": 5,
            "height": 200,
            "opacity": 0.2,
            "position": "absolute",
            "right": 150,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 510,
            "width": 200,
          },
          {
            "backgroundColor": "#1DB954",
            "shadowColor": "#1DB954",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 200,
            "bottom": -200,
            "elevation": 5,
            "height": 400,
            "opacity": 0.3,
            "position": "absolute",
            "right": -100,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "width": 400,
          },
          {
            "backgroundColor": "#1DB954",
            "shadowColor": "#1DB954",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 50,
            "bottom": 280,
            "elevation": 5,
            "height": 100,
            "opacity": 0.1,
            "position": "absolute",
            "right": 90,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "width": 100,
          },
          {
            "backgroundColor": "#1DB954",
            "shadowColor": "#1DB954",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 100,
            "elevation": 5,
            "height": 200,
            "left": 30,
            "opacity": 0.4,
            "position": "absolute",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 140,
            "width": 200,
          },
          {
            "backgroundColor": "rgba(10, 10, 10, 1)",
            "shadowColor": "rgba(10, 10, 10, 1)",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 75,
            "elevation": 5,
            "height": 150,
            "opacity": 0.4,
            "position": "absolute",
            "right": 40,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 150,
            "width": 150,
          },
          {
            "backgroundColor": "rgba(10, 10, 10, 1)",
            "shadowColor": "rgba(10, 10, 10, 1)",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 25,
            "elevation": 5,
            "height": 50,
            "opacity": 0.2,
            "position": "absolute",
            "right": 150,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 120,
            "width": 50,
          },
          {
            "backgroundColor": "#B026FF",
            "shadowColor": "#B026FF",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 125,
            "elevation": 5,
            "height": 250,
            "opacity": 0.4,
            "position": "absolute",
            "right": -60,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": -60,
            "width": 250,
          },
          {
            "backgroundColor": "#B026FF",
            "shadowColor": "#B026FF",
          },
        ]
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(10, 10, 10, 0.5)",
        "flexDirection": "row",
        "height": "13%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "height": 50,
          "left": 20,
          "opacity": 1,
          "position": "absolute",
          "top": "50%",
          "width": 50,
        }
      }
    >
      <Text />
    </View>
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 24,
          "maxWidth": "70%",
          "position": "absolute",
          "top": "50%",
        }
      }
    >
      Enter Room Code
    </Text>
  </View>
  <Text
    style={
      {
        "color": "#BBB",
        "fontSize": 22,
        "marginBottom": 10,
        "marginHorizontal": 60,
        "marginTop": 220,
      }
    }
  >
    Enter the code for the room you want to join
  </Text>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginVertical": 100,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 200,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "black",
            "borderColor": "grey",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 60,
            "marginHorizontal": 5,
            "width": 60,
          }
        }
      >
        <TextInput
          autoFocus={true}
          keyboardType="numeric"
          maxLength={1}
          onChangeText={[Function]}
          onKeyPress={[Function]}
          style={
            {
              "color": "white",
              "flex": 1,
              "fontSize": 24,
              "textAlign": "center",
            }
          }
          value=""
        />
      </View>
      <View
        style={
          {
            "backgroundColor": "black",
            "borderColor": "grey",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 60,
            "marginHorizontal": 5,
            "width": 60,
          }
        }
      >
        <TextInput
          autoFocus={false}
          keyboardType="numeric"
          maxLength={1}
          onChangeText={[Function]}
          onKeyPress={[Function]}
          style={
            {
              "color": "white",
              "flex": 1,
              "fontSize": 24,
              "textAlign": "center",
            }
          }
          value=""
        />
      </View>
      <View
        style={
          {
            "backgroundColor": "black",
            "borderColor": "grey",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 60,
            "marginHorizontal": 5,
            "width": 60,
          }
        }
      >
        <TextInput
          autoFocus={false}
          keyboardType="numeric"
          maxLength={1}
          onChangeText={[Function]}
          onKeyPress={[Function]}
          style={
            {
              "color": "white",
              "flex": 1,
              "fontSize": 24,
              "textAlign": "center",
            }
          }
          value=""
        />
      </View>
      <View
        style={
          {
            "backgroundColor": "black",
            "borderColor": "grey",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 60,
            "marginHorizontal": 5,
            "width": 60,
          }
        }
      >
        <TextInput
          autoFocus={false}
          keyboardType="numeric"
          maxLength={1}
          onChangeText={[Function]}
          onKeyPress={[Function]}
          style={
            {
              "color": "white",
              "flex": 1,
              "fontSize": 24,
              "textAlign": "center",
            }
          }
          value=""
        />
      </View>
      <View
        style={
          {
            "backgroundColor": "black",
            "borderColor": "grey",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 60,
            "marginHorizontal": 5,
            "width": 60,
          }
        }
      >
        <TextInput
          autoFocus={false}
          keyboardType="numeric"
          maxLength={1}
          onChangeText={[Function]}
          onKeyPress={[Function]}
          style={
            {
              "color": "white",
              "flex": 1,
              "fontSize": 24,
              "textAlign": "center",
            }
          }
          value=""
        />
      </View>
    </View>
  </View>
</View>
`;
